### ISSUE TITLE
Reduce the memory usage for ACL rules in shadowsocks-rust.

### ISSUE THREAD

**Issue by madeye** [shadowsocks/shadowsocks-android#2627]
I observed a very large memory usage of shadowsocks-rust recently when ACL is enabled. Actually, this issue exists from the first 5.1 version. The memory usage easily reaches to 200M in several minutes after the startup.

```
 9969 u0_a446      20   0  12G 197M 2.3M S 18.0   2.6   0:25.90 libsslocal.so +
```

With ACL disabled, the maximum memory usage is about 30M, which is expected.

I suspected it's related to the regex we're using now. Maybe we should move to wildcard rules to solve this issue.

**Comment by madeye** [comment]
From the Rust doc here https://docs.rs/regex/1.3.1/regex/struct.RegexSetBuilder.html#method.dfa_size_limit, I noticed:

> Note that this is a per thread limit. There is no way to set a global limit. In particular, if a regex is used from multiple threads simultaneously, then each thread may use up to the number of bytes specified here.

It looks that Rust's regex would duplicate DFA for each thread.

After enabling the `single-threaded` feature, I found the memory usage dropped dramatically. Now, the max memory usage is about 50M, much closer to what I observed before for shadowsocks-libev with PCRE (30M).
